### PR TITLE
Allow position tracking on point layers

### PR DESCRIPTION
**Pull request by nirvn**
This PR unlocks position tracking for point layers (implements https://github.com/opengisch/QField/issues/1921).
image

Contrary to line of polygon layers, tracking against a point layer will create one feature per triggered tracking record. This allows for all @position_* variables exposed here (https://qfield.org/docs/prepare/gnss.html) to be used in attributes' default value expression.

When starting tracking on a point layer, a feature form will pop up. Attribute values entered that _do not_ have a default value expression (or provider-side default value like FIDs) will be remembered across the created features. This means that for e.g., an attribute could be used to identify a user name creating a given point track, and that attribute value will be remembered during the tracking session. 

In terms of on-canvas display, when tracking on a point layer, QField will still display a line rubberband connecting all the points, which gives a nice visual feedback of the path on top of individual points. Pro tip: use a datetime attribute with a default value expression set to `now()`, that will allow you order the collected points and create paths from those points. The spatio-temporal ST-DBSCAN clustering algorithm can also re-group those points into different sessions by clustering with a max. duration interval.

Additional improvements and cleanups have been applied along the way, including:
- I've removed the arbitrary prohibition on constraint-less tracking. When neither time nor distance constraints are on, the tracking will record one vertex/point per position received by the internal or external GNSS device.
- A nice little UX boost: if the vector layer used to track has an empty feature form (i.e. all attributes are hidden), we now just skip showing an empty feature form and instead start tracking right away. That's a nice usability improvement for people working with users who might be intimidated by a feature form there.

_(The first commit has a bunch of code formatting and variable renaming to try and help readability for newcomers, I apologize for the resulting diff nightmare.)_

_Funded by [QGIS usergroup Switzerland](https://www.qgis.ch/en), send your thanks to them! :)_

### CODE REVIEW

**Comment by qfield-fairy** [comment]
🎉 Ta-daaa, freshly created APKs are available for e8520843e815968cbc720c805757e40cc9c41120: [**arm64-android**](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-3247-c7d294960b636a772d634a3f2829cbc59be08519-arm64-android.apk)
<details>
<summary>Other architectures</summary>

- [arm-android](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-3247-c7d294960b636a772d634a3f2829cbc59be08519-arm-android.apk)
- [x64-android](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-3247-c7d294960b636a772d634a3f2829cbc59be08519-x64-android.apk)
- [x86-android](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-3247-c7d294960b636a772d634a3f2829cbc59be08519-x86-android.apk)
</details>

**Comment by nirvn** [comment]
I've added a label to explain the frequency of tracking when both time and distance constraints are disabled:
image

**Comment by nirvn** [comment]
@m-kuhn , @suricactus , if you guys have a minute to review this one, I'd like to merge so I can refresh screenshots and documentation.

**Comment by nirvn** [comment]
Relevant documentation improvements: https://github.com/opengisch/QField-docs/pull/239

File: src/core/featuremodel.cpp [hunk 1]
```diff
@@ -472,7 +472,22 @@ void FeatureModel::resetFeature()
...
 
-void FeatureModel::resetAttributes()
+void FeatureModel::resetFeatureId()
+{
+  if ( !mLayer )
+    return;
+
+  if ( sRememberings->contains( mLayer ) )
+  {
+    QMutex *mutex = sMutex;
```
Comment (m-kuhn) [inline src/core/featuremodel.cpp:482]: What's this copy for?

Comment (nirvn) [reply src/core/featuremodel.cpp:482]: @m-kuhn , the logic I had in mind is that whenever you call either resetFeature() or the newly introduced resetFeatureId(), the attribute values held in the mFeature at the time of those calls are remembered. It's not used here per say, but I think if we rely on this new function elsewhere, we'd likely want the remember logic to match that of resetFeature().

Comment (m-kuhn) [reply src/core/featuremodel.cpp:482]: I meant the mutex

Comment (nirvn) [reply src/core/featuremodel.cpp:482]: Oh sorry I thought you meant to code block -- you know what, I coded that forever ago and completely forgot why I did it this way throughout featuremodel.cpp :fearful:

**Review by marioba — APPROVED** [review]
Looks good to me!

Great job @nirvn

**Pull request merged**